File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
         Node* curr, *next = NULL;
        curr = head;
        
        while(curr != NULL)
        {
            next = curr->next;  //preserve next
            Node *newNode = new Node(curr->val);
            newNode->next = curr->next; //link new nodes next with latter node
            curr->next = newNode;   //link it with new 
            curr = next;    //move
        }
        
        //linking on random pointers
        for(curr = head; curr != NULL;curr = curr->next->next)
        {
            //newNode random
            curr->next->random = (curr->random != NULL)?curr->random->next:NULL;
        }
        
        //separating both lists and relinking each of there nodes in respective list
        curr = head;
        Node* copy;
        Node* psuedoHead = new Node(0);
        copy = psuedoHead;  ///deep copy list
        
        while(curr != NULL)
        {
            
            next = curr->next->next;    //preserve next of original
            
            copy->next = curr->next;    //deep copy linking and its separation
            
            curr->next = next; //orignal's next
            
            //move
            copy = copy->next;
            curr = next;
        }
        
        return psuedoHead->next;
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
        def cloneInsert(self, head):
            ##step 1
            temp = head
            while temp is not None:
                copyNode = Node(temp.val)
                copyNode.next = temp.next
                temp.next = copyNode
                #move itr
                temp = temp.next.next
 
        def cloneRandomPointers(self, head):
            ##step 2
            temp = head
            while temp is not None:
                copyNode = temp.next  # start of cloned list
                if temp.random:
                    copyNode.random = temp.random.next
                # move itr
                temp = temp.next.next
        
        def segregateLinks(self, head):
            ##step 3
            dummyNode = Node(-1)
            res = dummyNode
            temp = head
            while temp is not None:
                res.next = temp.next  # start of cloned list
                # move itr
                temp.next = temp.next.next

                res = res.next
                temp = temp.next
            return dummyNode.next

        def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
            self.cloneInsert(head)
            self.cloneRandomPointers(head)
            return self.segregateLinks(head)
                
        